Fix bug: MPR executes tests multiple times

mjesun opened the pull request:
When executing a test by path (not sure why this is a thing?) we were returning `true` for each call in a project configured. This is because as soon as a test is a file, we returned `true`, not checking if the file actually belongs to the project being tested.

This has popped up multiple times (e.g. #5332).

thymikee commented on packages/jest-cli/src/search_source.js:
```diff
@@ -209,14 +209,22 @@ export default class SearchSource {
           }
+
+          // It exists and it is a file; return true if it's in the project.
+          return allFiles.includes(path.resolve(name));
```
@mjesun have you measured perf impact of this? I suspect `allFiles` to be quite a big array, no?

mjesun replied:
I can change it to a `Set` so that the search becomes O(1).

thymikee replied:
I mean, for the typical use-case (`jest some-path` or `jest test1 test2`) it should be negligible, but hypothetically it might hit the user harder when they provide a glob with tens of paths.

I'm good with a set.

mjesun commented:
* I've added a test which mimics the observed behavior (tests passed by path get executed N-times, once per config). I've also tested that the test fails on master and passes on the current branch.

* Added entry to the changelog pointing to this PR.

* Modified the code for checking if the file belongs to the project, by using a `Set` instead of the `Array` given by `hasteFS.getAllFiles()`.

cpojer commented on packages/jest-cli/src/search_source.js:
```diff
@@ -209,14 +209,22 @@ export default class SearchSource {
           try {
-            return fs.lstatSync(name).isFile();
+            if (!fs.lstatSync(name).isFile()) {
+              // It exists, but it is not a file; return false.
```
can you remove the "return false" here? It's literally the next line.

thymikee replied:
huh, I didn't spot that :D
